### PR TITLE
EnvVar to disable creation of default ingress controller

### PR DESCRIPTION
**Pull request by davidvossel**
Due to the way Hypershift is designed, it needs the ability to start the ingress operator in the hosted cluster namespace, and then create the default ingresscontroller object later on which hypershift's controllers manage. Right now we have a race between the ingress operator attempting to autocreate the default ingresscontroller and hypershift's controllers attempting to do it.

This PR adds an environment variable option to the operator to disable the autocreation of the default ingresscontroller which allows hypershift to manage the default ingresscontroller without conflict.

### CODE REVIEW

File: pkg/operator/operator.go
Comment (gcs278) [inline pkg/operator/operator.go:53]: I think it would be more consistent to add the `os.Getenv` to start.go and add this variable as a part of the `operatorconfig.Config` object in config/config.go. That'd at least keeps it aligned with other configuration options. Cluster-dns-operator uses ENV variables exclusively, and it uses the `operatorconfig.Config` pattern: https://github.com/openshift/cluster-dns-operator/blob/master/cmd/dns-operator/main.go#L23-L48 so I'm using that as a guide.

File: pkg/operator/operator.go
Comment (gcs278) [inline pkg/operator/operator.go:54]: *nit* this might just be me, but the verb IGNORE feels a bit awkward here. Would using DISABLE make more sense?

Also, the word "AUTO" implies to me that, we can still somehow manually tell the ingress-operator to manage the default ingress controller, but I don't think that's true. I think you are disabling the ingress-operator from managing the default ingress controller completely right?

Would something like `DISABLE_DEFAULT_INGRESS_CONTROLLER=true` be more simple and explicit?

File: pkg/operator/operator.go
Comment (gcs278) [inline pkg/operator/operator.go:229]: Regardless of where this logic ends up per my last comment about `OperatorConfig.config`, you should use `strconv.ParseBool(VAR)`

**Review by gcs278 — CHANGES_REQUESTED** [review]
Thanks for the PR, let me know if my comments make sense. I think I'd still like a more experienced member of our team consider the configuration split between ENVs and CMD line arguments, but I personally don't have an issue as I understand this PR is basically for Hypershift.

File: pkg/operator/operator.go
Comment (davidvossel) [reply pkg/operator/operator.go]: yep, makes sense

File: pkg/operator/operator.go
Comment (davidvossel) [reply pkg/operator/operator.go]: fixed

File: pkg/operator/operator.go
Comment (gcs278) [inline pkg/operator/operator.go:240]: One more thing. It'd probably be helpful to have a INFO level log here to say it's skipping creation of the default ingress controller.
`log.Info("skipping default ingress controller creation")` or something

**Review by gcs278 — COMMENTED** [review]
Looks good to me - just one more suggestion, but I still want a second opinion, probably from @Miciah on the pattern, the name of ENV variable, and the fact we have both CMD line args and ENV variables as configuration options.

File: pkg/operator/operator.go
Comment (davidvossel) [reply pkg/operator/operator.go:242]: updated

**Comment by davidvossel** [comment]
> Looks good to me - just one more suggestion, but I still want a second opinion, probably from @Miciah on the pattern, the name of ENV variable, and the fact we have both CMD line args and ENV variables as configuration options.

In our case, we'd prefer an ENV var over a cli option so we don't have to worry about supplying an unknown cli option to a cluster-ingress-operator that isn't yet aware of the option to disable default ingress controller.

With an ENV var, we can always set it in our hypershift logic, and the env var will just be ignored when the cluster-ingress-operator is unaware of the change.

**Comment by candita** [comment]
Has there been any kind of study or enhancement document on what might go wrong if the default ingress controller is disabled?  The CI tests are inconclusive so far.

**Comment by davidvossel** [comment]
> Could we instead check the control-plane topology in the infrastructure config (i.e., infraConfig.Status.ControlPlaneTopology == configv1.ExternalTopologyMode)?

I can't say with certainty this accurately represents when a default ingress operator needs to be present or not.

What needs to happen in order to move this PR forward?

**Comment by davidvossel** [comment]
> What needs to happen in order to move this PR forward?

specifically, what I'm looking for is a consensus on the direction to take here by the cluster ingress operator maintainers. I don't mind making the change to look at ExternalTopologyMode, or using the existing env var method. either will functionally work for me.

**Comment by Miciah** [comment]
> > Could we instead check the control-plane topology in the infrastructure config (i.e., infraConfig.Status.ControlPlaneTopology == configv1.ExternalTopologyMode)?
> 
> I can't say with certainty this accurately represents when a default ingress operator needs to be present or not.

Are there cases where Hypershift would want the ingress operator to create the default ingresscontroller?  Anyway, I thought `ExternalTopologyMode` was added specifically for Hypershift.  Maybe @csrwng could elucidate.  

> > What needs to happen in order to move this PR forward?
> 
> specifically, what I'm looking for is a consensus on the direction to take here by the cluster ingress operator maintainers. I don't mind making the change to look at ExternalTopologyMode, or using the existing env var method. either will functionally work for me.

I want to make sure we don't ignore established norms without good reason, and normally configuration is persisted in API objects, not injected using environment variables.  Is this approach of injecting environment variables the way Hypershift configures other components?  To @candita's point, is there a relevant enhancement document that describes Hypershift's use of environment variables to configure operators?

**Comment by davidvossel** [comment]
> Are there cases where Hypershift would want the ingress operator to create the default ingresscontroller?

no

> Anyway, I thought ExternalTopologyMode was added specifically for Hypershift. Maybe @csrwng could elucidate.

The value implies hypershift. 

I created #797 to disable the default ingress controller creation when ExternalTopologyMode is encountered. Any of these solutions will fix the race condition that exists today and I'm not partial to one over the other. Which ever one works the best for you all.

**Pull request closed without merging**
